Edit Shortcuts UI (dev/keyboardManager)

arjunbalgovind opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Added the Edit Shortcuts UI and reformatted the code to include a common library which has a state class that can be used to share information between the UI and the back-end.

## Details of the Pull Request
- Added a common library with Helper functions and KeyboardManagerState class. This class will be used to share information between the hook and the UI. The object of this class is created in dllmain.cpp.
- Refactored the code base to constrain all code which involves UI and back-end information to wrapper functions in the state class.
- Added a shortcutControl class which creates a custom control for each element in the shortcut table.

## Follow up To-Dos
* Add thread safety
* Add tests
* Add support to save the shortcuts to a file
* Add UI for app specific shortcuts and configurations
* Display key names instead of codes

<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist
* [X] Applies to #6 
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA
* [X] Tests passed

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
- Run PowerToys.exe. The Keyboard Manager window should pop up if PowerKeys is enabled.
- Click Edit Shortcuts. Clicking the + Sign will allow you to add a new shortcut. The shortcut can be removed by clicking the delete button.
- Press Type shortcut to detect your shortcut. As of now the key name will not be displayed, only the code.
- In the current version, only 2 key shortcuts will be applied. A check has not been added for the order of the keys in the UI, so while testing it make sure the modifier key is clicked first. Example: Ctrl+A and not A+Ctrl.
- Clicking Apply will save the settings and your shortcuts will be remapped. If any of remaps were invalid, the valid ones will still apply, however a warning will be displayed on the top right corner.
- Clicking Cancel will simply exit the window. If Apply was not clicked prior to clicking cancel then those shortcuts will not be saved.
- If you reopen the Edit Shortcuts Window your shortcuts will still be visible. However if you terminate PowerToys.exe and relaunch it your saved shortcuts will be lost.

![editshortcuts](https://user-images.githubusercontent.com/32061677/77211389-16aeac80-6ac1-11ea-9ae0-4d60669c1445.gif)

arjunbalgovind committed:
Added EditShortcuts Window and added Detecting shortcuts functionality

arjunbalgovind committed:
Fixed build error

arjunbalgovind committed:
Changed detection to take place only when window is in focus

arjunbalgovind committed:
Added solution folder

arjunbalgovind committed:
Added a common project and refactored shared variables to an object with wrapper functions

arjunbalgovind committed:
Added dynamic addition of shortcuts

arjunbalgovind committed:
Moved all shared variables in detection to state variable with wrapper functions

arjunbalgovind committed:
Added code to re-load saved shortcuts in the UI

arjunbalgovind committed:
Added comments

crutkas commented:
why not in settings just do what we do with FZ Editor where you click a button?

ryanbodrug-microsoft commented on src/modules/keyboardmanager/common/Helpers.cpp:
```diff
@@ -0,0 +1,26 @@
+#include "Helpers.h"
+
+// Function to split a wstring based on a delimiter and return a vector of split strings
+std::vector<std::wstring> splitwstring(std::wstring input, wchar_t delimiter)
```
`const std::wstring& input` would avoid the copy here.

arjunbalgovind commented:
> why not in settings just do what we do with FZ Editor where you click a button?

@crutkas That would be what will be done eventually when the stuff on the `MainWindow` is shifted to the settings page. For now since we want to keep it on WinUI, I structured it like this and it can be migrated with settings v2. The part from after clicking Edit Shortcuts is still independent of the settings screen.

ryanbodrug-microsoft commented on src/modules/keyboardmanager/common/Helpers.h:
```diff
@@ -0,0 +1,58 @@
+
+// Function to convert an unsigned int vector to hstring by concatenating them
+template<typename T>
+hstring convertVectorToHstring(std::vector<T>& input)
```
input should be const

ryanbodrug-microsoft commented on src/modules/keyboardmanager/common/Helpers.h:
```diff
@@ -0,0 +1,58 @@
+
+// Function to convert a wstring vector to a integer vector
+template<typename T>
+std::vector<T> convertWStringVectorToIntegerVector(std::vector<std::wstring> input)
```
`const std::vector<std::wstring>& input`

arjunbalgovind replied:
Will make changes like this to all the functions I've added in these files. I wasn't very aware on how the `const` on the argument works. A quick search showed that it would be super helpful to make sure the code isn't doing something horribly wrong :)

ryanbodrug-microsoft commented on src/modules/keyboardmanager/common/KeyboardManagerState.h:
```diff
@@ -0,0 +1,112 @@
+using namespace Windows::UI::Xaml::Controls;
+
+// Enum type to store different states of the UI
+enum class KeyboardManagerUIState
```
Nice use of scoped enum :)

arjunbalgovind committed:
Fixed argument modifiers in Helpers

ryanbodrug-microsoft commented on src/modules/keyboardmanager/common/Helpers.h:
```diff
@@ -0,0 +1,58 @@
+#pragma once
+#include <windows.h>
+#include <stdlib.h>
```
Do you really need all these includes in the header file?  If possible your better off moving the includes to the .cpp file and `forward declaring` the types referenced.

arjunbalgovind replied:
I've also noticed that having all these WinUI header files adds a lot to compilation time. Does that mean it would be better to move them to a pre-compiled header and include that directly on all the files?

ryanbodrug-microsoft replied:
Yep, these would probably be good candidates for a pch.  I wouldn't expect them to change very often.

ryanbodrug-microsoft commented on src/modules/keyboardmanager/common/KeyboardManagerState.h:
```diff
@@ -0,0 +1,112 @@
+#pragma once
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <map>
```
Duplicate comment:  I"m not sure you need *all* the includes in the header file.  This can slow down build times.

arjunbalgovind committed:
Updated arg modifiers in all functions

ryanbodrug-microsoft reviewed (commented):
The changes look good to me.  There are a few places where vectors and strings look like they could be pass by const & instead of by value.  I didn't label all of them.  Could you do a pass on the function signatures, and make sure they are a) 'const correct' meaning methods and parameters are const by default unless they can't be, and b) avoid unessary copies during pass by value.

arjunbalgovind committed:
Removed unused headers and added precompiled headers

arjunbalgovind commented:
> The changes look good to me.  There are a few places where vectors and strings look like they could be pass by const & instead of by value.  I didn't label all of them.  Could you do a pass on the function signatures, and make sure they are a) 'const correct' meaning methods and parameters are const by default unless they can't be, and b) avoid unessary copies during pass by value.

Resolved all the function signatures, removed unused headers and moved WinUI headers to precompiled header files.

ryanbodrug-microsoft reviewed (approved):
Nice work!
